Define and implement QueryKey Cursors

EasterTheBunny opened the pull request:
Cursors in the EVM implementation of QueryKey are constructed to match the current default sorting in LogPoller using block number and log index. Using both of these pieces of data together uniquely identifies a log within the LogPoller database.

ilija42 commented on core/chains/evm/logpoller/parser.go:
```diff
@@ -268,15 +268,14 @@ func (v *pgDSLParser) whereClause(expressions []query.Expression, limiter query.
 			return "", errors.New("invalid cursor direction")
 		}
 
-		block, txHash, logIdx, err := valuesFromCursor(limiter.Limit.Cursor)
```
in case of reorg this is not an unique event identifier, txHash_logIdx pair is the only evm unique identifier

EasterTheBunny replied:
That is true, good callout. If we leave `txHash` in the cursor, it wouldn't alter the order in any way for log order since we still have `logIdx`. In that case, we won't need to make any changes at all.

reductionista replied:
> in case of reorg this is not an unique event identifier, txHash_logIdx pair is the only evm unique identifier

Maybe I should look at the surrounding code, but are you sure this is something we need?  LogPoller always returns blocks from the current canonical chain, it can never return more than one block with the same block number in a given query. In any given query, block number + log index uniquely identifies each log and uniquely specifies the order of them.  (In fact, it's not even possible to store multiple rows in the db for the same (block_number, log_index) combination)... those are always guaranteed to be unique.  The only case where it might be something you need is if something wants to hang on to old blocks that are re-org'd out and keep track of both them and the canonical blocks separately.  But usually I would think you'd *want* them to be identified with the same key so that you can replace the old one with the current one.  (Why would a product ever care about old blocks that have been re-org'd out, unless it's just to find and replace them?)

ilija42 replied:
Even if log poller handles reorgs this is more precise and helpful if products need to log this value, I don't see a reason not to do this

reductionista replied:
> If we leave txHash in the cursor, it wouldn't alter the order in any way for log order since we still have logIdx

Depends what order you have those listed in.  If it's the same order as the return values above (block, txHash, logIdx) then leaving txHash in there will definitely destroy the ordering.  The transaction hashes are random, so it will make everything in random order instead of being in order of how the logs were emitted on chain.

If you really need to include txHash for some reason, then you must list it after logIdx so that it doesn't ruin the ordering. But I have a hard time imaging having it in there would be useful for anything.

reductionista replied:
> Even if log poller handles reorgs this is more precise and helpful if products need to log this value, I don't see a reason not to do this

Don't they already get that field back with each log?  I wouldn't say it's more precise, just more verbose... logIdx is more precise in that it tells you about the order rather than just adding an unrelated piece of information (what the SHA is, associated with the transaction it's from)

ilija42 replied:
Cursor present in the log event should have txHash and since that is the same cursor getting passed into here they should be the same

EasterTheBunny replied:
Resolved by moving `txHash` to the end of the cursor and limiting cursor usage to finalized.

EasterTheBunny committed:
Define and implement QueryKey Cursors

Cursors in the EVM implementation of QueryKey are constructed to match the current
default sorting in LogPoller using block number and log index. Using both of these
pieces of data together uniquely identifies a log within the LogPoller database.

ilija42 commented on core/chains/evm/logpoller/parser.go:
```diff
@@ -268,15 +272,14 @@ func (v *pgDSLParser) whereClause(expressions []query.Expression, limiter query.
+		segment = fmt.Sprintf("%s AND block_number %s= :cursor_block AND log_index %s :cursor_log_index", segment, op, op)
 
 		v.args.withField("cursor_block_number", block).
-			withField("cursor_txhash", common.HexToHash(txHash)).
```
Isn't it misleading if we are ignoring txHash from the cursor? 

If we handle it like this then this can happen:
- Reorg happened which changed the txHash but log poller handled it somehow and logs are returned(which shouldn't ever happen since we are limiting this to finalized logs)
- No reorg, all good

Similar cases if we do filter logs by the txHash:
- Reorg happened, logPoller handled it, but we return an error here since txHash doesn't match
- Reorg happened, logPoller didn't handle it and the caller gets an error ( or log poller is in degraded state which is basically the same for the caller)
- No reorg, all good

Is there still a discussion to be had here on how to handle this? I am not a fan of passing in txHash and then not using it. That said this might also be fine, but I would leave some comments around this code to remind us of this

EasterTheBunny replied:
I thought the idea was that we didn't need to be concerned with reorgs because we aren't allowing cursors to be used with anything other than finalized.

ilija42 replied:
Yes, but why not also give extra info and be more precise?

EasterTheBunny replied:
The `txHash` is already in the cursor, so the amount of info doesn't change, per our conversation. Adding the `txHash` to the query behind the scenes becomes an irrelevant no-op when limiting to finalized blocks and won't affect precision.

reductionista replied:
> Yes, but why not also give extra info and be more precise?

I think I may have misunderstood what @ilija42 was suggesting in our last meeting. I am okay with returning tx hash as a result, as attaching extra irrelevant info can't hurt anything.  But I'm not okay with using it to restrict the queries, as that would lead to unpredictable wrong results... making it easy to miss logs.

reductionista replied:
> Isn't it misleading if we are ignoring txHash from the cursor?

I agree that it's counter-intuitive... ideally, we should not be passing it to the cursor as in the best case scenario it would not affect anything and in the worse case it would do actual harm... in no case would it be able to improve the results as far as I can see.

My first choice would still be to not use it for anything cursor-related.  But if we are going to use it, my 2nd choice would be to just return it as a result but not accept it as an input parameter.  A distant 3rd choice would be passing it as an input parameter and not using it (or only using it for ORDER BY which is guaranteed by the db constraints to be ignored).  While using it for anything that might have an effect (like filtering on it) I think is unacceptable.

EasterTheBunny committed:
add tx hash to SQL query

reductionista commented on core/chains/evm/logpoller/parser_test.go:
```diff
@@ -46,16 +47,19 @@ func TestDSLParser(t *testing.T) {
+			"AND block_number <= " +
+			"(SELECT finalized_block_number FROM evm.log_poller_blocks WHERE evm_chain_id = :evm_chain_id ORDER BY block_number DESC LIMIT 1)) " +
+			"AND block_number >= :cursor_block AND tx_hash >= :tx_hash AND log_index > :cursor_log_index " +
+			"ORDER BY block_number ASC, log_index ASC, tx_hash ASC " +
```
It's harmless to return the `tx_hash` as a result, or to include it at the end of an ORDER BY (since the PRIMARY KEY on the table is (block_hash, log_index, evm_chain_id), which guarantees the tx_hash part of the ORDER BY will be ignored due to it being impossible for that to have any effect).  But I don't think we should include `tx_hash` in the AND clause like this, as that could return the wrong results if it did ever matter.

In principle, it shouldn't matter, since finalized logs are not supposed to ever change after they're written to the table.  But if that assumption were violated then there's a 50/50 chance you'd get the wrong result back, since the hash is entirely random.  Doesn't that defeat the only rationale for ius doing anything with the tx_hash in the first place? Instead of improving the robustness in a hypothetical edge case, it makes the results more unpredictable and error-prone.

Consider the following scenario:

1. QueryKey is called, asking for the most recent logs of a particular type in ascending order.
It returns 20 logs with the most recent having block_number=10, log_index=5, tx_hash=0x84.
2. A re-org happens (or backup logpoller runs, updating an old log), so that block is now replaced with a whole new set of logs. The log in the new block with log_index=4 still has the same block_number=5, but now it has a tx hash that happens to begin with 0x13.
3. QueryKey is called again, using CursorLimit("10-5-0x42", CursorFollowing, 20) to get any logs more recent than that back.
4. Because through random chance, the hash of the new transaction is less than the hash of the old one, the new one is not returned in the query.

This makes me realize something else also... why are we using >= for block_number and tx_hash but > for log_index?

reductionista replied:
Actually, come to think of it.. there are worse problems with this query. Including tx_hash here will cause it to return the wrong results even in the typical case, not just in a hypothetical scenario where a finalized log gets updated after it's marked finalized.

Combining these range restrictions with AND doesn't work here.  The restriction `tx_hash > :tx_hash` filters out any results whose hashes don't happen to satisfy that criteria regardless of what block they are in or what their log_index is.  So even if you had, say 100 important log results in block 11, saying block_number >= 10 AND tx_hash >= 0x80 will filter out approximately 50 from the results you should have gotten in block 11!

And the problem is bigger than that, it's not just with tx_hash the same problem can happen with log_index.  If the starting point of your cursor has a high log_index, this will never give you any results with a lower log_index even if they are in a more recent block.

reductionista replied:
I think we need the WHERE clause to end up looking like this:
```
block_number > :block_number OR (block_number = :block_number AND log_index > :log_index)
```

EasterTheBunny replied:
I updated the query to match the `OR` condition you created above.

EasterTheBunny committed:
modify cursor query for accurate block and log querying

reductionista commented on core/chains/evm/logpoller/parser.go:
```diff
@@ -272,15 +272,14 @@ func (v *pgDSLParser) whereClause(expressions []query.Expression, limiter query.
 		}
 
-		block, logIdx, txHash, err := valuesFromCursor(limiter.Limit.Cursor)
+		block, logIdx, _, err := valuesFromCursor(limiter.Limit.Cursor)
```
I have a feeling most products are going to use _ here to ignore txhash also.  I still can't think of any potential use a product would have for getting a txhash back here... it just doesn't seem relevant to the feature they are using.

I think the PR as it stands now will give correct results, so I'm willing to approve it if everyone else is happy.  But just noting that I still think it would make a lot more sense to not involve txhash at all in this cursor feature... it's harmless, but doesn't add any value either IMO
